### PR TITLE
Update The missing piece towards any scale geospatial analytics.md

### PR DESCRIPTION
**Pull request by PahudPlus**
Editorial updates.

### CODE REVIEW

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md [hunk 1]
```diff
@@ -1,48 +1,48 @@
-# The missing piece towards 'any scale' spatial-temporal analytics
+# The missing piece in 'any scale' spatial-temporal analytics
```
Comment (yxm1536) [inline 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: 这里用in会改变原意。原意为“为了能够做到any scale，我们现在缺少那几块拼图”。
如果没有语法问题，towards的表达会更为恰当

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md [hunk 2]
```diff
@@ -1,48 +1,48 @@
...
 
-The Internet of Things (IoT) and 5G technologies will involve an astounding amount of data that record the links between people, device, event, location and time. According to IDC's forecast, there will be 79.4 zettabytes of data generated by 41.6 billion IoT devices in 2025. With the rapid growth of IoT data and IoT applications, there is an ever rising demand for a highly efficient spatial-temporal data science workflow to gain insights from the overwhelming data.
+Internet of Things (IoT) and 5G technologies will involve an astounding amount of data interconnecting people, device, event, location, and time. According to IDC's forecast, there will be 79.4 zettabytes of data generated by 41.6 billion IoT devices in 2025. With the rapid growth of IoT data and IoT applications, there is an ever rising demand for a highly efficient spatial-temporal data science workflow to gain insights from the overwhelming data.
```
Comment (yxm1536) [inline 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: 讲 data 将 people, device等互联起来，主谓搭配感觉很奇怪。原文表达的意思是数据记录了这些关系。

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md [hunk 3]
```diff
@@ -1,48 +1,48 @@
...
 
-Scalability is the key to build productive data science pipelines. To address the scalability challenge, we launch Arctern, an open source spatial-temporal analytic framework for boosting end-to-end data science performance. Arctern aims to improve scalability from two aspects: 1. unified data analytic and processing interface across different platforms, from laptops to clusters and cloud; 2. rich and consistent algorithms and models, including trajectory processing, spatial clustering and regression, etc., across different data science pipeline stages. 
+Scalability is key to building productive data science pipelines. To address this challenge, we launched Arctern, an open source spatial-temporal analytic framework for boosting end-to-end data science performance. Arctern aims to improve scalability from two aspects: 1. Unified data analytic and processing interface across different platforms, from laptops to clusters and cloud; 2. Rich and consistent algorithms and models, including trajectory processing, spatial clustering and regression, etc., across different data science pipeline stages.
```
Comment (yxm1536) [inline 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: this challenge 这里用 the scalability challenge会更明确。
因为前面的Scalability is the key to build productive data science pipelines. 并没有太明确的表达出哪个是challenge

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md [hunk 4]
```diff
@@ -1,48 +1,48 @@
...
+To better understand the scalability problem existing in the current workflows, we need to take a look at the pipeline of spatial-temporal data science, which is illustrated in the figure above. Raw data is first generated by IoT devices, and then collected by data store. Data scientists make exploratory analysis over these data. Basing on technical hypothesis and data features, they select proper models to develop a prototype for answering business questions. After a few iterations of evaluation and adjustment, the model will finally be deployed in a data processing system, which in turn delivers better services to end users via IoT devices.
 
-The whole process can be further divided into three stages:
+The whole process can break down to three stages:
```
Comment (yxm1536) [inline 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md:15]: 请确认break down不需要是被动

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md [hunk 5]
```diff
@@ -1,48 +1,48 @@
...
+* *Prototype development*. In this stage, data scientists tend to work with small datasets for the purpose of fast data exploration, efficient model selection, and iterative prototype polishing. Lightweight tools are favored in this stage, because they can provide quick insights and support fast iteration. Typical tools include NumPy, GeoPandas, scikit-learn, PySAL, and kepler.gl.
+* *Model evaluation*. To fit the prototype into the production environment, much larger (or even close to production level) datasets are used for model evaluation. As a result, the model evaluation-modification iterations in this stage are much less agile than in the prototype stage. Moreover, the tool chain is much heavier. Most of the commonly-used tools, such as PostGIS and PySpark, are disk-based, or distributed.
+* *Model deployment*. Deployment is the stage, where the model is integrated into the existing production environment to provide practical services. This stage involves large scale systems such as HDFS, Hbase, TSDB for storage, and Spark, Flink, GeoMesa for data processing.
```
Comment (yxm1536) [inline 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: Deployment is the stage, where the model is integrated ....
这里stage后面的逗号能否去掉，读起来这里卡一下有点难受

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md [hunk 6]
```diff
@@ -1,48 +1,48 @@
...
 
 ### The scalability problems
 
-At different stages of spatial-temporal data science, the demand for tool chain varies significantly. Although libraries and systems mentioned above can jointly support the overall data science workflow, each individual tool can hardly scale across multiple data science stages or across different analytic platforms (laptop, cluster, cloud). As a result, the tool-box grows too heavy such that the complexity makes it difficult to organize the workflow in an highly efficient manner. Specifically, the challenges are three-fold:
+At different stages of spatial-temporal data science, demand for tool chain varies significantly. Although libraries and systems mentioned above can jointly support the overall data science workflow, an individual tool can hardly scale across multiple data science stages or across different analytic platforms (laptop, cluster, or cloud). As a result, the tool chain grows increasingly complex to a point where it is hard to streamline the workflow. Specifically speaking, the challenges are three-pronged:
```
Comment (yxm1536) [inline 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: three-pronged 这里应该用 three-fold。

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md [hunk 7]
```diff
@@ -1,48 +1,48 @@
...
+* *The gap between library (system) interfaces*. Libraries or systems used in different stages do not share a unified interface. As a consequence, data scientists may use multiple languages such as Python, R, Scala, SQL and inconsistent function names throughout the process of data analysis, prototype development, model evaluation, and model deployment. Such interface inconsistency significantly steepens the learning curve, and forces data scientists to spend valuable time and brainpower rewriting their codes, rather than focusing on business questions.
+* *The gap between algorithm/model libraries*. Libraries used in different stages use different algorithms and models. The agreement of standardization is now still at a primitive level, as the ones proposed by the Open Geospatial Consortium (OGC). This raises great problems on productivity. When an advanced algorithm is used as part of the solution during prototype development, but no proper implementation is found for the production environment, great efforts have to be made to reinvent the wheel.
```
Comment (yxm1536) [inline 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: 这段话理解歧义。首先不同的数据科学阶段是会使用相同的算法和模型的，只是各个阶段所用的工具对于算法和模型的支持程度不同。其次primitive这里指原语，而不是初级。
这段还是直接用之前的版本吧。调整起来比较麻烦

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md [hunk 8]
```diff
@@ -1,48 +1,48 @@
...
+* *The gap between data scientist and engineer*. Spatial-temporal data science is a complex work in which mathematical problems, technical problems and business problems need to be properly handled. It calls for seamless collaboration between data scientists and engineers. However, the inconsistency of the tools used throughout the workflow poses a burden during communication. Given that different programming languages and algorithm implementations are used in different data science stages, the engineers may have no idea as to which models the data scientists used, whilst the data scientists are struggling to understand why the engineers cannot make their models work.
```
Comment (yxm1536) [inline 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: which models 这里要用 what models

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md [hunk 9]
```diff
@@ -1,48 +1,48 @@
...
 
-In the next few sections, we will compare several libraries and systems widely used in spatial-temporal data science, and give more discussions on the missing piece towards 'any scale' spatial-temporal analytics. 
+In the next few sections, we will compare several libraries and systems widely used in spatial-temporal data science, and give more discussion on the missing piece in 'any scale' spatial-temporal analytics.
```
Comment (yxm1536) [inline 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: in vs. towards。同上

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md [hunk 10]
```diff
@@ -1,48 +1,48 @@
...
+These tools, albeit popular, cannot scale to crunch big data. They were inherently designed for small datasets on a standalone machine. Considering hardware capacity of modern laptops and workstations, the size of data that these tools can handle would be up to a couple of GBs. The figure below shows the performance of 19 frequently-used primitives in GeoPandas. As we can observe, when data size reaches 10M, most of the primitives in GeoPandas take several minutes to complete. PySAL shares the same problem with GeoPandas when handling large datasets. Compared with the primitives in GeoPandas, the advanced algorithms in PySAL generally have higher computation complexity, thus may cause even severer efficiency problems on data analysis and prototype development. Since data scientists need to iteratively adjust and verify their models until they get satisfying results, these primitives and algorithms are frequently and repeatedly used. The scalability problem would seriously degrade efficiency in the spatial data science workflow.
```
Comment (yxm1536) [inline 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: albeit popular这个用法感觉比较少见

Comment (yxm1536) [inline 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: standalone machine这个讲法不常见。建议还是用single

Comment (yxm1536) [reply 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: Considering hardware capacity of modern laptops and workstations, the size of data that these tools can handle would be up to a couple of GBs. 去掉hardware
=>
Considering the capacity of modern laptops and workstations, the size of data that these tools can handle would be up to a couple of GBs.

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md [hunk 11]
```diff
@@ -1,48 +1,48 @@
...
 
 ![](geopandas_test.png)
 
-Due to the scalability problem, the above mentioned tools are generally used in prototype design with small dataset. To work with large production data, data scientists and engineers would eventually need to apply their algorithms in distributed solutions, such as GeoSpark and GeoMesa.
+Due to the scalability problem, the above mentioned tools are generally used in prototype design with small dataset. To work with large production-level datasets, data scientists and engineers would eventually need to apply their algorithms to distributed solutions, such as GeoSpark and GeoMesa.
```
Comment (yxm1536) [inline 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: would eventually need to apply their algorithms to distributed solutions, such as GeoSpark and GeoMesa. apply 改成migrate
=>
would eventually need to migrate their algorithms to distributed solutions, such as GeoSpark and GeoMesa.

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md [hunk 12]
```diff
@@ -52,7 +52,7 @@ GeoSpark is a cluster computing system that extends Apache Spark and SparkSQL wi
 | GeoSpark | GeoSpark Spatial RDD | SparkSQL, java, scala |
 | GeoMesa Spark | GeoMesa Spatial RDD | SparkSQL, java, scala |
 
-Compared with GeoPandas and PySAL, GeoSpark and GeoMesa adopt very different API abstraction. The Table above presents an interface comparison. The data interfaces vary greatly among different tools, with no pair of tools sharing identical data interface. In terms of programming language, there are three major categories: python, SQL, and Java/Scala. 
+Compared with GeoPandas and PySAL, GeoSpark and GeoMesa adopt very different API abstraction. The Table above presents an interface comparison. The data interfaces vary greatly with tools, and no pair of tools share the same data interface. In terms of programming language, there are three major categories: python, SQL, and Java/Scala.
```
Comment (yxm1536) [inline 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: 这段建议用原版。

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md [hunk 13]
```diff
@@ -63,32 +63,32 @@ Compared with GeoPandas and PySAL, GeoSpark and GeoMesa adopt very different API
 | GeoSpark | 37 | spatial join, etc. |
 | GeoMesa Spark | 75 | spatial join, KNN search, proximity search, route search, tube selection, etc. |
 
-The table above compares the provided primitives and algorithms among geospatial analysis tools. The number of supported OGC standard primitives shows large difference, ranging from around 30 to three hundreds. The advanced algorithms show great divergence as well. Each of the tools offers a limited set of algorithms that barely overlap.
+The table above compares the provided primitives and algorithms among geospatial analysis tools. The number of supported OGC standard primitives shows large difference, ranging from around 30 to three hundred. The advanced algorithms show great divergence as well. Each of the tools offers a limited set of algorithms that do not overlap.
```
Comment (yxm1536) [inline 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: do not 建议用 barely，感觉意思不一样。

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md [hunk 14]
```diff
@@ -63,32 +63,32 @@ Compared with GeoPandas and PySAL, GeoSpark and GeoMesa adopt very different API
...
+In Arctern, to help data scientists and engineers improve productivity and creativity, we try to fix the inconsistency problem of spatial data science tools by providing rich primitives, algorithms, and models with consistent interfaces across laptop, cluster and cloud environments. 
 
-Among the tools we studied, we found the interfaces of GeoPandas well fits the usage habits of data scientist in their workflow, meanwhile is easy to use. Therefore, we adopt its idea and plan to build the GeoDataFrame/GeoSeries that scale both up and out. Upon which we will further develop a consistent spatial-temporal algorithm set across execution environments.
+Among the tools we studied, we found that GeoPandas has interfaces working best with the usage habits of data scientist and is easy to use. Therefore, we adopt its idea and plan to build the GeoDataFrame/GeoSeries that scale both up and out. On top of this, we will develop a consistent spatial-temporal algorithm set across execution environments.
```
Comment (yxm1536) [inline 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md:74]: interfaces working best with habits 这个感觉很奇怪

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md [hunk 15]
```diff
@@ -63,32 +63,32 @@ Compared with GeoPandas and PySAL, GeoSpark and GeoMesa adopt very different API
...
 
-At present, we have developed an efficient multi-thread GeoSeries implementation, and the distributed version is in progress. In the latest version 0.2.0, Arctern achieves 24x speed up against GeoPandas. Even under single-thread execution, Arctern outperforms GeoPandas 7x on average. The detailed evaluation results is illustrated in the figure below.
+We have now developed an efficient multi-thread GeoSeries implementation, and the distributed version is in progress. In the latest version 0.2.0, Arctern achieves 24x speed up against GeoPandas. Even under single-thread execution, Arctern outperforms GeoPandas 7x on average. The detailed evaluation results are illustrated in the figure below.
 
 ![](geoseries_speedup.png)
 
-In terms of interaction/exploration on large datasets, Arctern provides rendering methods that can render data (as heatmap, choropleth, etc.) with a volume of 10M in seconds. 
+In terms of interaction/exploration on large datasets, Arctern provides rendering methods that can render data (such as heatmap, choropleth, etc.) with a volume of 10M in seconds.
```
Comment (yxm1536) [inline 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: 原版表达的意思是把数据渲染成heatmap, choropleth, etc
Arctern provides rendering methods that can render data , as heatmap, choropleth, etc., with a volume of 10M in seconds.

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md [hunk 16]
```diff
@@ -63,32 +63,32 @@ Compared with GeoPandas and PySAL, GeoSpark and GeoMesa adopt very different API
...
 * Enriching our spatial-temporal algorithm sets. We will concentrate on KNN search and trajectory analysis in the project's early stages. 
 
 ### About the Authors
 
-Rentong Guo is a Director of R&D at ZILLIZ. He received the Ph.D. degree in Computer Software and Theory from Huazhong University of Science and Technology (HUST), China. He spends his career participating in and leading works focused on database, cache system, distributed system, heterogeneous computing.
+Rentong Guo is a Director of R&D at ZILLIZ. He received a Ph.D. degree in Computer Software and Theory from Huazhong University of Science and Technology (HUST), China. He spends his career participating in and leading projects in the area of database, cache system, distributed system, and heterogeneous computing.
```
Comment (yxm1536) [inline 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md:91]: He received a Ph.D. degree 
=> 
He received his Ph.D. degree

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md [hunk 17]
```diff
@@ -63,32 +63,32 @@ Compared with GeoPandas and PySAL, GeoSpark and GeoMesa adopt very different API
...
 
 
-Xiaomeng Yi, Senior researcher at ZILLIZ. Xiaomeng Yi received his Ph.D. degree in Computer Architecture from Huazhong University of Science and Technology. His research focuses on scheduling and resource allocation in the cloud and distributed systems. His works are published in prestigious conferences and journals, such as IEEE Network Magazine, IEEE TON, IEEE ICDCS, and ACM TOMPECS.
\ No newline at end of file
+Xiaomeng Yi is a senior researcher at ZILLIZ. Xiaomeng Yi received his Ph.D. degree in Computer Architecture from Huazhong University of Science and Technology. His research focuses on scheduling and resource allocation in the cloud and distributed systems. His works have been published in prestigious conferences and journals, such as IEEE Network Magazine, IEEE TON, IEEE ICDCS, and ACM TOMPECS.
```
Comment (yxm1536) [inline 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: His works have been published
=>
His works are published

**Review by yxm1536 — CHANGES_REQUESTED** [review]

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md (hunk 1 above)
Comment (PahudPlus) [reply 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: of?  “towards” awkward to me.

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md (hunk 13 above)
Comment (PahudPlus) [reply 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: barely 是勉强不

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md (hunk 3 above)
Comment (PahudPlus) [reply 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: this scalability challenge.

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md (hunk 14 above)
Comment (PahudPlus) [reply 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md:74]: fitting well with

Comment (PahudPlus) [reply 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md:74]: fitting well with

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md (hunk 16 above)
Comment (PahudPlus) [reply 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md:91]: Either way is fine.

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md (hunk 2 above)
Comment (PahudPlus) [reply 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: recording the links between

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md (hunk 13 above)
Comment (PahudPlus) [reply 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: that hardly overlap

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md (hunk 6 above)
Comment (PahudPlus) [reply 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md:23]: come from three aspects.

File: 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md (hunk 10 above)
Comment (PahudPlus) [reply 2020_0602 The missing piece towards any scale geospatial analytics/The missing piece towards any scale geospatial analytics.md]: Despite their popularity. Either way is fine.

**Review by yxm1536 — APPROVED** [review]

**Pull request merged**